src/data/zha.ts

```diff
@@ -0,0 +1,117 @@
+import { HassEntity } from "home-assistant-js-websocket";
+import { HomeAssistant } from "../types";
+
+export interface ZHADeviceEntity extends HassEntity {
+  device_info?: {
+    identifiers: any[];
+  };
+}
+
+export interface ZHAEntities {
+  [key: string]: HassEntity[];
+}
+
+export interface Attribute {
+  name: string;
+  id: number;
+}
+
+export interface Cluster {
+  name: string;
+  id: number;
+  type: string;
+}
+
+export interface Command {
+  name: string;
+  id: number;
+  type: string;
+}
+
+export interface ReadAttributeServiceData {
+  type: string;
+  entity_id: string;
+  cluster_id: number;
+  cluster_type: string;
+  attribute: number;
+  manufacturer: number;
+}
+
+export const fetchAttributesForCluster = (
+  hass: HomeAssistant,
+  entityId: string,
+  ieeeAddress: string,
+  clusterId: number,
+  clusterType: string
+): Promise<Attribute[]> =>
+  hass.callWS({
+    type: "zha/entities/clusters/attributes",
+    entity_id: entityId,
+    ieee: ieeeAddress,
+    cluster_id: clusterId,
+    cluster_type: clusterType,
+  });
+
+export const computeReadAttributeServiceData = (
+  entityId: string,
+  clusterId: number,
+  clusterType: string,
+  attributeId: number,
+  manufacturerCodeOverride: any,
+  selectedEntity?: ZHADeviceEntity,
+  selectedCluster?: Cluster,
+  selectedNode?: HassEntity
+): ReadAttributeServiceData | undefined => {
+  if (!selectedEntity || !selectedCluster || !selectedNode) {
```

Comment:
We should not have this check inside here, instead this method should not be called if there is nothing selected.